### PR TITLE
PLAT-10261: Fixed classpath paths for windows platform

### PR DESCRIPTION
**Pull request by symphony-elias**
### Ticket
PLAT-10261

### Description
Fixed path joining for classpath resources on Windows.

### Checklist
- [x] Referenced a ticket in the PR title and in the corresponding section
- [x] Filled properly the description and dependencies, if any
- [x] Unit tests updated or added

### CODE REVIEW

File: symphony-bdk-legacy/symphony-api-client-java/src/test/java/internal/FileHelperTest.java [hunk 1]
```diff
@@ -29,4 +29,24 @@ public void should_read_file_from_system() throws IOException {
   public void fail_to_read_file() throws FileNotFoundException {
     FileHelper.readFile(UUID.randomUUID().toString() + ".abc");
   }
-}
\ No newline at end of file
+
```
Comment (symphony-youri) [inline symphony-bdk-legacy/symphony-api-client-java/src/test/java/internal/FileHelperTest.java:32]: may I suggest a test with just classpath:, /file and classpath:, file
it looks it could be a corner case too

for this kind of test that are very simple and in the form of one input, one output I would sometime have a single test method and all the test cases in the same method
(but I don't mind having separated methods too :))

File: symphony-bdk-legacy/symphony-api-client-java/src/main/java/internal/FileHelper.java [hunk 2]
```diff
@@ -28,30 +28,33 @@
...
-        }
-    }
-    else if (FileHelper.class.getResource(classpath(path)) != null) {
-      content = toByteArray(FileHelper.class.getResourceAsStream(classpath(path)));
+    if (!isClasspath(path) && new File(path).exists()) {
+      try (FileInputStream fis = new FileInputStream(path)) {
+        logger.debug("File loaded from system path : {}", path);
+        return toByteArray(fis);
+      } catch (Exception ex) {
+        throw new FileNotFoundException("Unable to load file from path : " + path);
```
Comment (symphony-youri) [inline symphony-bdk-legacy/symphony-api-client-java/src/main/java/internal/FileHelper.java:36]: just noted that there is an extra space before :

File: symphony-bdk-legacy/symphony-api-client-java/src/main/java/internal/FileHelper.java [hunk 3]
```diff
@@ -28,30 +28,33 @@
...
   }
 
-  public static String path(String first, String... more) {
-    return Paths.get(first, more).toString();
+  public static String path(String first, String second) {
+    return isClasspath(first) ? classpathPath(first, second) : Paths.get(first, second).toString();
+  }
+
+  private static String classpathPath(String first, String second) {
+    // classpath paths are always with '/'
```
Comment (symphony-youri) [inline symphony-bdk-legacy/symphony-api-client-java/src/main/java/internal/FileHelper.java:50]: are you sure? to me classpath:blah.txt is a valid classpath

File: symphony-bdk-legacy/symphony-api-client-java/src/main/java/internal/FileHelper.java [hunk 4]
```diff
@@ -28,30 +28,33 @@
...
-    else if (FileHelper.class.getResource(classpath(path)) != null) {
-      content = toByteArray(FileHelper.class.getResourceAsStream(classpath(path)));
+    if (!isClasspath(path) && new File(path).exists()) {
+      try (FileInputStream fis = new FileInputStream(path)) {
+        logger.debug("File loaded from system path : {}", path);
+        return toByteArray(fis);
+      } catch (Exception ex) {
+        throw new FileNotFoundException("Unable to load file from path : " + path);
+      }
+    } else if (FileHelper.class.getResource(classpath(path)) != null) {
```
Comment (symphony-youri) [inline symphony-bdk-legacy/symphony-api-client-java/src/main/java/internal/FileHelper.java:38]: it is a bit weird that here we use getResource first and then getResourceAsStream (it was like that before)

also to load relative from this class is perhaps a bit weird for the user of the SDKs
perhaps we should just use ClassLoader.getSystemResourceAsStream ? (I started looking into ClassPathResource from Spring and it gets messy quickly, picking the right classloader, forward slash or not...)

File: symphony-bdk-legacy/symphony-api-client-java/src/main/java/internal/FileHelper.java (hunk 3 above)
Comment (symphony-elias) [reply symphony-bdk-legacy/symphony-api-client-java/src/main/java/internal/FileHelper.java:50]: I didn't want to say classpath paths always begin with '/' but rather that it cannot contain '\' like Windows (path separator is always '/' regardless of the OS).

File: symphony-bdk-legacy/symphony-api-client-java/src/main/java/internal/FileHelper.java (hunk 4 above)
Comment (symphony-elias) [reply symphony-bdk-legacy/symphony-api-client-java/src/main/java/internal/FileHelper.java:38]: Yes indeed, in Spring things are messy.
But please check: https://docs.oracle.com/javase/8/docs/technotes/guides/lang/resources.html

> **Using Methods of java.lang.Class**
The getResource and getResourceAsStream methods find a resource with a given name. They return null if they do not find a resource with the specified name. The rules for searching for resources associated with a given class are implemented by the class's ClassLoader. The Class methods delegate to ClassLoader methods, after applying a naming convention: if the resource name starts with "/", it is used as is. Otherwise, the name of the package is prepended, after converting all periods (.) to slashes (/).

So for our usecases, I guess it is enough and will work if paths begin with `classpath:/`. Things get messier if there is no slash after `classpath:`.

File: symphony-bdk-legacy/symphony-api-client-java/src/test/java/internal/FileHelperTest.java [hunk 5]
```diff
@@ -29,4 +30,40 @@ public void should_read_file_from_system() throws IOException {
...
+  }
+
+  @Test
+  public void only_classpath_on_first_slash_on_second() {
+    assertEquals("classpath:/file.pem", FileHelper.path("classpath:", "/file.pem"));
+  }
+
+  @Test
+  public void only_classpath_first_with_no_slash() {
+    assertEquals("classpath:/file.pem", FileHelper.path("classpath:", "file.pem"));
```
Comment (symphony-elias) [inline symphony-bdk-legacy/symphony-api-client-java/src/test/java/internal/FileHelperTest.java:66]: Please mind that `Paths.get("classpath:", "file.pem")` (=current behavior) will add a path separator in between (`/` on linux/macOS or `\` on Windows)

**Review by symphony-youri — APPROVED** [review]

File: symphony-bdk-legacy/symphony-api-client-java/src/main/java/internal/FileHelper.java [hunk 6]
```diff
@@ -28,30 +28,31 @@
...
-      try(FileInputStream fis = new FileInputStream(path)) {
-          content = toByteArray(fis);
-          logger.debug("File loaded from system path : {}", path);
-        }
-        catch (Exception ex) {
-          throw new FileNotFoundException("Unable to load file from path : " + path);
-        }
+  private static String classpathPath(String first, String second) {
+    // Path separator for classpath paths is always '/' regardless of the OS
+    if (first.endsWith("/") && second.startsWith("/")) {
```
Comment (symphony-youri) [inline symphony-bdk-legacy/symphony-api-client-java/src/main/java/internal/FileHelper.java:49]: maybe have a constant for the / ?
